machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,131 @@
+
+	// spec describes the configuration of the machine os build
+	// +kubebuilder:validation:Required
+	Spec MachineOSBuildSpec `json:"spec"`
+
+	// status describes the lst observed state of this machine os build
+	// +optional
+	Status MachineOSBuildStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineOSObjectReference `json:"machineConfigPool"`
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSObjectReference `json:"renderedMachineOSImage"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig string `json:"currentConfig"`
+	// desiredConfig is the desired config we want to build an image for. If currentConfig and desiredConfig are not the same, we need to build an image.
+	// +kubebuilder:validation:Required
+	DesiredConfig string `json:"desiredConfig"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// buildStart describes when this build began
+	// +kubebuilder:validation:Required
+	BuildStartTime metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended
+	// +optional
+	BuildEndTime metav1.Time `json:"buildEnd"`
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Built
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// relatedBuilds contains previous iterations of failed or interrupted builds related to this one.
+	// This will keep track of failed or interrupted build names
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	RelatedBuilds []PriorMachineOSBuilds `json:"relatedBuilds,omitempty"`
```

Comment:
Maybe I need to read the enhancement, but, perhaps take a look at the Job construct in K8s and see how they handle tracking history, if at all.

Would I not expect to be able to `oc get machineosbuild` and see the historical items?

Reply:
this part is still being figured out. I am leaning towards this flow:

1. one MachineOSBuild item per pool or group of nodes using the same config
2. this object exists for the lifetime that rendered-pool-abc configs exist and need to be managed
3. this field exists for the purpose of tracking "Failed" or "Interrupted" builds

I was originally thinking each BUILD gets its own MachineOSBuild object. This could make sense as well? thoughts @cheesesashimi ?

Reply:
How does a pool opt in to an on cluster builds?